src/main/scala/com/mozilla/telemetry/learning/federated/FederatedLearningSearchOptimizerS3Sink.scala

```diff
@@ -0,0 +1,151 @@
+}
+
+class FederatedLearningSearchOptimizerS3Sink(outputPath: String, stateCheckpointPath: String) extends Sink {
+
+  private[federated] var state: OptimisationState = initState()
+
+  override def addBatch(batchId: Long, data: DataFrame): Unit = {
+    val aggregates = data.collect().map(FrecencyUpdateAggregate(_))
+
+    val iteration: Long = state.iteration
+
+    val ord = Ordering.by((_: FrecencyUpdateAggregate).modelVersion)
+    aggregates.filter(_.studyVariation == "treatment").filter(_.modelVersion.toLong >= iteration).reduceOption(ord.min) match {
+      case None =>
+      // no relevant updates, do nothing
+      case Some(aggregate) =>
+
+        val lastWeights: Array[Double] = state.weights
+        val learningRates: Array[Double] = state.learningRates
+        val previousGradient: Option[Array[Double]] = state.gradient
+
+        val gradient = aggregate.avgUpdates
+        val InternalResult(newWeights, newLearningRates) = FederatedLearningRPropOptimizer.fit(lastWeights, gradient, previousGradient, learningRates)
+
+        val newIteration = iteration + 1
+
+        // val log = Array(OptimisationLogLine(aggregate.window.start, aggregate.modelVersion.toLong, aggregate.avgLoss))
+
+        val newState = OptimisationState(newIteration, newWeights, newLearningRates, Option(gradient))
+        writeModel(ModelOutput(newWeights.map(math.round(_).toInt), newIteration))
+
+        writeState(newState)
+        state = newState
+    }
+  }
+
+  private[federated] def writeModel(modelOutput: ModelOutput): Unit = {
+    implicit val formats = DefaultFormats
+    val jsonModel = Serialization.write(modelOutput)
+
+    val conf = SparkHadoopUtil.get.conf
+    val modelOutputPath = new Path(outputPath)
+
+    val fs = modelOutputPath.getFileSystem(conf)
+    if (!fs.exists(modelOutputPath)) {
+      fs.mkdirs(modelOutputPath)
+    }
+
+
+    val latestStream = fs.create(new Path(outputPath + "/" + "latest.json"))
+    latestStream.writeBytes(jsonModel)
+    latestStream.close()
+
+    val versionedStream = fs.create(new Path(outputPath + "/" + modelOutput.iteration + ".json"))
+    versionedStream.writeBytes(jsonModel)
+    versionedStream.close()
+
+    fs.close()
+  }
+
+  private[federated] def writeState(state: OptimisationState): Unit = {
+    implicit val formats = DefaultFormats
+    val jsonState = Serialization.write(state)
+
+    val fileName = "STATE-" + state.iteration
+    val path = new Path(stateCheckpointPath + "/" + fileName)
+
+    val fs = path.getFileSystem(SparkHadoopUtil.get.conf)
+    val outputStream = fs.create(path)
+    outputStream.writeBytes(jsonState)
+
+    outputStream.close()
+    fs.close()
+  }
+
+  private[federated] def initState(): OptimisationState = {
+    val conf = SparkHadoopUtil.get.conf
+    val checkpointPath = new Path(stateCheckpointPath)
+    val fs = checkpointPath.getFileSystem(conf)
+    if (!fs.exists(checkpointPath)) {
+      fs.mkdirs(checkpointPath)
```

Comment:
Hmm, how does this work for S3, re: keys not existing if there's no file inside

Reply:
It works :)